psi4/src/psi4/libfock/DFJCOSK.cc

```diff
@@ -0,0 +1,1012 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along
+ * with Psi4; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * @END LICENSE
+ */
+
+#include "jk.h"
+#include "psi4/libqt/qt.h"
+#include "psi4/libfock/cubature.h"
+#include "psi4/libfock/points.h"
+#include "psi4/libmints/vector.h"
+#include "psi4/libmints/basisset.h"
+#include "psi4/libmints/electrostatic.h"
+#include "psi4/libmints/mintshelper.h"
+#include "psi4/libmints/molecule.h"
+#include "psi4/libmints/integral.h"
+#include "psi4/liboptions/liboptions.h"
+#include "psi4/lib3index/dftensor.h"
+
+#include <vector>
+#include <map>
+#include <algorithm>
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
+using namespace psi;
+
+namespace psi {
+
+Matrix compute_numeric_overlap(const std::shared_ptr<DFTGrid> &grid, const std::shared_ptr<BasisSet> &primary) {
+
+    // DOI 10.1063/1.3646921, EQ. 9
+
+    int nbf = primary->nbf();
+    BasisFunctions bf_computer(primary, grid->max_points(), grid->max_functions());
+    Matrix S_num("Numerical Overlap", nbf, nbf);
+    auto S_nump = S_num.pointer();
+
+    // This loop could be parallelized over blocks of grid points. However, the cost of the loop is
+    // so small (< 10 seconds for a 200 heavy atom system), parallelism isn't necessary
+
+    for (const auto &block : grid->blocks()) {
+
+        // grid points in this block
+        int npoints_block = block->npoints();
+        int nbf_block = block->local_nbf();
+        auto w = block->w();
+
+        // compute basis functions at these grid points
+        bf_computer.compute_functions(block);
+        auto point_values = bf_computer.basis_values()["PHI"];
+
+        // resize the buffer of basis function values
+        Matrix X_block("phi_g,u", npoints_block, nbf_block);  // points x nbf_block
+        auto X_blockp = X_block.pointer();
+        for (size_t p = 0; p < npoints_block; p++) {
+            for (size_t k = 0; k < nbf_block; k++) {
+                X_blockp[p][k] = point_values->get(p, k) * std::sqrt(w[p]);
+            }
+        }
+
+        // significant basis functions at these grid points
+        const auto &bf_map = block->functions_local_to_global();
+
+        auto S_num_block = linalg::doublet(X_block, X_block, true, false);
+        auto S_num_blockp = S_num_block.pointer();
+
+        for (size_t mu_local = 0; mu_local < nbf_block; mu_local++) {
+            size_t mu = bf_map[mu_local];
+            for (size_t nu_local = 0; nu_local < nbf_block; nu_local++) {
+                size_t nu = bf_map[nu_local];
+                S_nump[mu][nu] += S_num_blockp[mu_local][nu_local];
+            }
+        }
+
+    }
+
+    S_num.hermitivitize();
+
+    return S_num;
+
+}
+
+SharedMatrix compute_esp_bound(std::shared_ptr<BasisSet> primary) {
```

Comment:
More unnecessary pointers! Return `Matrix` and take in `const BasisSet&`.

Reply:
I can certainly change the return type to `Matrix`, that makes sense.

The basis set is already stored as a `std::shared_ptr<BasisSet> primary_` on the JK class. To me, it seems cleaner to simply pass the already existing `shared_ptr`, rather than dereference it and pass it as a `const BasisSet&`. What's the reasoning for the latter?

Reply:
I prefer references to pointers when at all possible. As written, the function _only_ works for a `shared_ptr`. If I need to compute the bound with a `BasisSet`, `unique_ptr<BasisSet`, or `* BasisSet`, I can't use this function without some kind of extra work. While it's true that you need some extra work to de-reference it, it's as simple as passing in `*varname` rather than `varname`.

The other advantage is that we can now enforce `const` on the incoming `BasisSet` easily.